oauth_tools/external_idp.py

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import abc
+import logging
+from os.path import join
+from time import sleep
+from typing import List, Optional
+
+import requests
+from lightkube import Client, KubeConfig, codecs
+from lightkube.core.exceptions import ApiError, ObjectDeleted
+from lightkube.resources.apps_v1 import Deployment
+from lightkube.resources.core_v1 import Namespace, Pod, Service
+from pytest_operator.plugin import OpsTest
+from requests.exceptions import RequestException
+
+from oauth_tools.constants import (
+    DEX_CLIENT_ID,
+    DEX_CLIENT_SECRET,
+    DEX_MANIFESTS,
+    EXTERNAL_USER_EMAIL,
+    EXTERNAL_USER_PASSWORD,
+    KUBECONFIG,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalIdpManager(abc.ABC):
```

Comment:
Just a **NONBLOCKING** question (probably somehow pedantic, feel free to ignore): is it better to just name it `ExternalIdpService`? Because literally, a manager does not need to know and expose the attributes of its managed service from what I understand, and this class does not do much service management work.

If we would like to go with manager naming, I guess only abstracting the lifecycle management behaviors is good enough.

Reply:
I agree, I will change the name